packages/puppeteer-core/src/common/bidi/Frame.ts

```diff
@@ -207,4 +209,19 @@ export class BidiFrame extends Frame {
     this.sandboxes[MAIN_SANDBOX][Symbol.dispose]();
     this.sandboxes[PUPPETEER_SANDBOX][Symbol.dispose]();
   }
+
+  #exposedFunctions = new Map<string, ExposeableFunction<never[], unknown>>();
+  override async exposeFunction<Args extends unknown[], Ret>(
+    name: string,
+    apply: (...args: Args) => Awaitable<Ret>
+  ): Promise<void> {
+    if (this.#exposedFunctions.has(name)) {
+      throw new Error(
+        `Failed to add page binding with name ${name}: globalThis['${name}'] already exists!`
+      );
+    }
+    const exposeable = new ExposeableFunction(this, name, apply);
+    await exposeable.expose();
+    this.#exposedFunctions.set(name, exposeable);
```

Comment:
we should probably set it before we await `.expose` to avoid races

Reply:
If `expose` fails, it's not exposed then, so we shouldn't set it.

Reply:
imo, then we need to remove the failed function. The current version does not work as a mutex

Reply:
Indeed!